content/backend/graphql-scala/4-deferred_resolvers.md

```diff
@@ -0,0 +1,168 @@
+---
+title: Deferred Resolvers
+pageTitle: "Usage of Deferred Resolvers for better performance"
+description: "In this chapter, at first you will learn how to use deferred resolvers for better performance."
+question: "Check the correct statement about Deferred Resolvers"
+answers: ["Fetcher is higher level API for Deferred Resolver", "Deferred Resolver is of optimizing a query, Fetcher is for taking data from data source in batch","Fetcher is light version of Deferred Resolver without cache or support for relations", "Deferred Resolver and Fetcher are completely different, not related one to each other things" ]
+correctAnswer: 0
+---
+
+
+### Deferred Resolvers and Fetchers
+
+`Fetcher`s and `Deferred Resolver`s are mechanisms for batch retrieval of object from their sources like database or external API. `Deferred Resolver` provides low-level efficient API but, on the other hand, it’s more complicated in use and less secured. `Fetcher` is a specialized version of `Deferred Resolver`. It provides high-level API, it’s easier to use and usually provides all you need. It optimizes resolution of fetched entities based on its ID or relation, it deduplicates entities and caches the results, to name but a few possibilities.
+
+Let's implement one for `Link`:
+
+<Instruction>
+
+Add definition of fetcher for link to the `GraphQLSchema` object:
+
+```scala
+
+val linksFetcher = Fetcher(
+  (ctx: MyContext, ids: Seq[Int]) => ctx.dao.getLinks(ids)
+)
+
+```
+
+</Instruction>
+
+Now we have to change fields in `QueryType`:
+
+<Instruction>
+
+Change Fields definition for the `link` and `links` with the following code:
+
+```scala
+
+Field("link",
+  OptionType(LinkType),
+  arguments = Id :: Nil,
+  resolve = c => linksFetcher.deferOpt(c.arg[Int]("id"))
+),
+Field("links",
+  ListType(LinkType),
+  arguments = List(Argument("ids", ListInputType(IntType))),
+  resolve = c => linksFetcher.deferSeq(c.arg[Seq[Int]]("ids"))
+)
+
+```
+
+</Instruction>
+
+We're still using `dao.getLinks` to fetch links from database, but now it's wrapped in `Fetcher`. It optimizes the query **before** call. Firstly it gathers all data it should fetch and then it execuees the queries. Caching and deduplication mechanisms allow to avoid duplicated queries and give results faster.
+
+ok, if we have our fetcher defined, we push it to lower level.
+
+<Instruction>
+
+in `GraphQLSchema` create constant for deferred resolver.
+
+```scala
+
+val Resolver = DeferredResolver.fetchers(linksFetcher)
```

Comment:
Lacks imports and for some reason IDEA broke for me ... at one point there might be a problem with the DAO - we used `Links` instead of `DBSchema.Links`